edb/server/pgcon/pgcon.pyx

```diff
@@ -116,39 +118,141 @@ def _build_init_con_script() -> bytes:
     ''').encode('utf-8')
 
 
-async def connect(connargs, dbname):
-    global INIT_CON_SCRIPT
+def _set_tcp_keepalive(transport):
+    # TCP keepalive was initially added here for special cases where idle
+    # connections are dropped silently on GitHub Action running test suite
+    # against AWS RDS. We are keeping the TCP keepalive for generic
+    # Postgres connections as the kernel overhead is considered low, and
+    # in certain cases it does save us some reconnection time.
+    sock = transport.get_extra_info('socket')
+    sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
+    if hasattr(socket, 'TCP_KEEPIDLE'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPIDLE, TCP_KEEPIDLE)
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPINTVL, TCP_KEEPINTVL)
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPCNT, TCP_KEEPCNT)
+
+
+async def _create_ssl_connection(protocol_factory, host, port, *,
+                                 loop, ssl_context, ssl_is_advisory):
+    tr, pr = await loop.create_connection(
+        lambda: TLSUpgradeProto(loop, host, port,
+                                ssl_context, ssl_is_advisory),
+        host, port)
+    _set_tcp_keepalive(tr)
+
+    tr.write(struct.pack('!ll', 8, 80877103))  # SSLRequest message.
+
+    try:
+        do_ssl_upgrade = await pr.on_data
+    except (Exception, asyncio.CancelledError):
+        tr.close()
+        raise
+
+    if do_ssl_upgrade:
+        try:
+            new_tr = await loop.start_tls(
+                tr, pr, ssl_context, server_hostname=host)
+        except (Exception, asyncio.CancelledError):
+            tr.close()
+            raise
+    else:
+        new_tr = tr
+
+    pg_proto = protocol_factory()
+    pg_proto.is_ssl = do_ssl_upgrade
+    pg_proto.connection_made(new_tr)
+    new_tr.set_protocol(pg_proto)
+
+    return new_tr, pg_proto
+
+
+class _Retry(Exception):
```

Comment:
I'd rename to `_RetryWithSSLSignal` and add `_RetryWithoutSSLSignal`

Reply:
Hm, having two exceptions would complicate things.  In this implementation both the [no-ssl, ssl] and [ssl, no-ssl] are served by the same piece of code which doesn't actually care if a particular set of connection args has ssl or not.

Reply:
Then `_RetryConnectSignal`